File: frontend/src/lib/asset-assessment.js
// Descriptive product diagnostics, separate from expected-return rankings.
// Inputs must carry a source, observation date and an explicit unit.
export const ASSESSMENT_VERSION = '1.1.0';
import specs from './product-policy.json' with { type: 'json' };
const rounded = x => Math.floor(x * 10 + .5 + 1e-9) / 10;
const day = s => typeof s === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(s) && Number.isFinite(Date.parse(s)) && new Date(s).toISOString().slice(0, 10) === s ? Date.parse(s) / 86400000 : null;
export function productAssessment(stock) {
  const dates = [day(stock.scoring?.priceAsOf), day(stock.productDataAsOf)].filter(d => d != null);
  const reference = dates.length ? Math.max(...dates) : null, fields = {}, missing = [], reasons = {};
  for (const [key, spec] of Object.entries(specs)) {
    const raw = stock.productMetrics?.[key];
    const entry = raw && typeof raw === 'object' && !Array.isArray(raw) && Object.keys(raw).length ? raw : null;
    const observed = day(entry?.asOf);
    const available = day(entry?.availableAt ?? entry?.asOf), expiry = day(entry?.validThrough);
    const valid = reference != null && observed != null && reference >= observed && reference - observed <= spec.maxAge
      && available != null && available <= reference && (!entry?.validThrough || (expiry != null && reference <= expiry))
      && typeof entry?.value === 'number' && Number.isFinite(entry.value) && entry.value >= 0 && entry.value <= spec.limit
      && entry.unit === spec.unit && typeof entry.source === 'string' && /^https:\/\//.test(entry.source)
      && (key !== 'dailyTrackingError' || (entry.method === 'sample_std_daily_difference' && entry.fundBasis === 'nav_total_return' && (!stock.targetReturnBasis || entry.benchmarkBasis === stock.targetReturnBasis) && Number.isFinite(entry.observations) && entry.observations >= 20 && Number.isFinite(stock.leverageMultiple) && !!stock.benchmark && entry.targetMultiple === stock.leverageMultiple && entry.benchmark === stock.benchmark));
    const index = valid ? spec.thresholds.findIndex(t => entry.value <= t) : -1;
    fields[key] = valid ? { ...entry, score: [95, 80, 60, 40, 20][index < 0 ? 4 : index] } : null;
    if (!valid) {
      missing.push(key);
      reasons[key] = !entry ? 'missing' : expiry != null && reference > expiry ? 'expired' : observed != null && reference - observed > spec.maxAge ? 'stale' : available != null && reference < available ? 'not_yet_available' : 'invalid';
    }
  }
  // All three execution dimensions are mandatory; AUM and concentration are not substitutes.
  const score = missing.length ? null : rounded(Object.entries(specs).reduce((sum, [key, spec]) => sum + fields[key].score * spec.weight, 0));
  return { score, asOf: reference == null ? null : new Date(reference * 86400000).toISOString().slice(0, 10), reasons, coverage: rounded(Object.entries(specs).reduce((sum, [key, spec]) => sum + (fields[key] ? spec.weight * 100 : 0), 0)), fields, missing };
}

export function leverageScenarios(multiple) {
  if (!Number.isFinite(multiple) || multiple === 1 || multiple === 0 || Math.abs(multiple) > 5) return [];
  return ['up', 'down', 'round_trip'].flatMap(path => [1, 5, 20].map(days => {
    let benchmark = 1, product = 1;
    for (let i = 0; i < days; i++) {
      const r = path === 'up' ? .01 : path === 'down' ? -.01 : i % 2 ? 1 / 1.01 - 1 : .01;
      benchmark *= 1 + r; product *= 1 + multiple * r;
    }
    return { path, days, benchmarkReturn: rounded((benchmark - 1) * 100), productReturn: rounded((product - 1) * 100) };
  }));
}

export function attachAssessments(stocks) {
  for (const stock of stocks) {
    if (stock.assetType === 'stock') { delete stock.assetAssessment; continue; }
    const candidates = stocks.filter(s => s.ticker === stock.underlyingSymbol && s !== stock && s.assetType === 'stock');
    const underlying = stock.underlyingType === 'stock' && candidates.length === 1 ? candidates[0] : null;
    stock.assetAssessment = {
      version: ASSESSMENT_VERSION, mode: stock.assetType === 'crypto' ? 'crypto_research' : 'product_research',
      product: stock.isETF ? productAssessment(stock) : null,
      underlying: underlying ? { ticker: underlying.ticker, qualityScore: underlying.qualityScore, timingScore: underlying.timingScore, priceAsOf: underlying.scoring.priceAsOf } : null,
      scenarios: stock.isETF ? leverageScenarios(stock.leverageMultiple) : [],
      cryptoCategory: stock.assetType === 'crypto' ? (['monetary', 'network', 'application', 'stablecoin'].includes(stock.cryptoCategory) ? stock.cryptoCategory : 'unknown') : null,
    };
    if (stock.productDataAsOf && stock.isETF) {
      const fee = stock.assetAssessment.product.fields.expenseRatio;
      stock.expenseRatio = fee?.value ?? null;
      stock.expenseRatioAsOf = fee?.asOf ?? null;
      // Legacy volatility estimates have no verified split/return basis; do not display them as current drag.
      stock.decayRate = null;
    }
    // Product quality and price trend have different meanings: never synthesize a buy score.
    stock.score = stock.qualityScore = stock.rank = stock.scoreSmoothed = stock.scoreDelta5d = stock.scoreHistoryVersion = null;
    stock.scoring.compositeEligible = false;
    stock.scoring.status = 'separate_dimensions';
    stock.scoring.qualityCoverage = 0;
    stock.scoring.coverage = stock.assetAssessment.product?.coverage ?? 0;
    stock.scoring.qualityDeduction = 0;
    stock.scoring.warnings = stock.scoring.warnings.filter(w => !w.includes('规模为流动性代理') && !w.includes('该资产类型尚未启用'));
    for (const key of ['cost', 'liquidity', 'diversification']) if (key in stock.subScores) stock.subScores[key] = null;
  }
  return stocks;
}

File: backend/asset_assessment.py
"""Product execution diagnostics. Never combine them with expected-return scores."""
import json
import math
from datetime import date
from pathlib import Path

SPECS = json.loads((Path(__file__).resolve().parents[1] / 'frontend/src/lib/product-policy.json').read_text())


def rounded(value):
    return math.floor(value * 10 + .5 + 1e-9) / 10


def day(value):
    try:
        parsed = date.fromisoformat(value)
        return parsed.toordinal() if parsed.isoformat() == value else None
    except (ValueError, TypeError):
        return None


def product_assessment(stock):
    dates = [d for d in (day(stock.get('scoring', {}).get('priceAsOf')), day(stock.get('productDataAsOf'))) if d is not None]
    reference = day(stock['evaluationAsOf']) if stock.get('evaluationAsOf') else max(dates) if dates else None
    fields, missing, reasons = {}, [], {}
    metrics = stock.get('productMetrics')
    if not isinstance(metrics, dict):
        metrics = {}
    for key, spec in SPECS.items():
        entry = metrics.get(key)
        if not isinstance(entry, dict):
            entry = {}
        observed, value = day(entry.get('asOf')), entry.get('value')
        available, expiry = day(entry.get('availableAt', entry.get('asOf'))), day(entry.get('validThrough'))
        valid = (reference is not None and observed is not None and 0 <= reference - observed <= spec['maxAge']
                 and available is not None and available <= reference and (not entry.get('validThrough') or (expiry is not None and reference <= expiry))
                 and isinstance(value, (int, float)) and not isinstance(value, bool) and math.isfinite(value)
                 and 0 <= value <= spec['limit'] and entry.get('unit') == spec['unit']
                 and isinstance(entry.get('source'), str) and entry['source'].startswith('https://')
                 and (key != 'dailyTrackingError' or (entry.get('method') == 'sample_std_daily_difference' and entry.get('fundBasis') == 'nav_total_return'
                      and (not stock.get('targetReturnBasis') or entry.get('benchmarkBasis') == stock['targetReturnBasis'])
                      and isinstance(entry.get('observations'), (int, float))
                      and math.isfinite(entry['observations']) and entry['observations'] >= 20 and isinstance(stock.get('leverageMultiple'), (int, float)) and bool(stock.get('benchmark')) and entry.get('targetMultiple') == stock.get('leverageMultiple')
                      and entry.get('benchmark') == stock.get('benchmark'))))
        score = next((v for threshold, v in zip(spec['thresholds'], [95, 80, 60, 40], strict=True) if value <= threshold), 20) if valid else None
        fields[key] = {**entry, 'score': score} if valid else None
        if not valid:
            missing.append(key)
            reasons[key] = ('missing' if not entry else 'expired' if expiry is not None and reference is not None and reference > expiry else
                            'stale' if observed is not None and reference is not None and reference - observed > spec['maxAge'] else
                            'not_yet_available' if available is not None and reference is not None and reference < available else 'invalid')
    score = None if missing else rounded(sum(fields[k]['score'] * spec['weight'] for k, spec in SPECS.items()))
    return {'score': score, 'asOf': date.fromordinal(reference).isoformat() if reference is not None else None, 'reasons': reasons, 'coverage': rounded(sum(spec['weight'] * 100 for k, spec in SPECS.items() if fields[k])), 'fields': fields, 'missing': missing}


def leverage_scenarios(multiple):
    if not isinstance(multiple, (int, float)) or not math.isfinite(multiple) or multiple in (0, 1) or abs(multiple) > 5:
        return []
    result = []
    for path in ('up', 'down', 'round_trip'):
        for days in (1, 5, 20):
            benchmark = product = 1
            for i in range(days):
                r = .01 if path == 'up' else -.01 if path == 'down' else 1 / 1.01 - 1 if i % 2 else .01
                benchmark *= 1 + r
                product *= 1 + multiple * r
            result.append({'path': path, 'days': days, 'benchmarkReturn': rounded((benchmark - 1) * 100), 'productReturn': rounded((product - 1) * 100)})
    return result


def attach_assessments(stocks):
    for stock in stocks:
        if stock['assetType'] == 'stock':
            stock.pop('assetAssessment', None)
            continue
        candidates = [s for s in stocks if s['ticker'] == stock.get('underlyingSymbol') and s is not stock and s['assetType'] == 'stock']
        underlying = candidates[0] if stock.get('underlyingType') == 'stock' and len(candidates) == 1 else None
        category = stock.get('cryptoCategory')
        stock['assetAssessment'] = {
            'version': '1.1.0', 'mode': 'crypto_research' if stock['assetType'] == 'crypto' else 'product_research',
            'product': product_assessment(stock) if stock.get('isETF') else None,
            'underlying': {k: underlying.get(k) for k in ('ticker', 'qualityScore', 'timingScore')} | {'priceAsOf': underlying['scoring']['priceAsOf']} if underlying else None,
            'scenarios': leverage_scenarios(stock['leverageMultiple']) if stock.get('isETF') else [],
            'cryptoCategory': (category if category in ('monetary', 'network', 'application', 'stablecoin') else 'unknown') if stock['assetType'] == 'crypto' else None,
        }
        if stock.get('productDataAsOf') and stock.get('isETF'):
            fee = stock['assetAssessment']['product']['fields']['expenseRatio'] or {}
            stock['expenseRatio'] = fee.get('value')
            stock['expenseRatioAsOf'] = fee.get('asOf')
            stock['decayRate'] = None
        for key in ('score', 'qualityScore', 'rank', 'scoreSmoothed', 'scoreDelta5d', 'scoreHistoryVersion'):
            stock[key] = None
        stock['scoring'].update(compositeEligible=False, status='separate_dimensions', qualityCoverage=0,
                                coverage=(stock['assetAssessment']['product'] or {}).get('coverage', 0), qualityDeduction=0)
        stock['scoring']['warnings'] = [w for w in stock['scoring']['warnings'] if '规模为流动性代理' not in w and '该资产类型尚未启用' not in w]
        for key in ('cost', 'liquidity', 'diversification'):
            if key in stock['subScores']:
                stock['subScores'][key] = None
